Feature/90 disambiguate first index of

ColmBhandal opened the pull request:
## Description

Refactored various indexing functions on the IEnumerable extension and the OneBasedArray class to disambiguate between zero-based and one-based indexing.

ColmBhandal committed:
Added failing test to capture problem scenario

ColmBhandal committed:
Added stub disambiguation methods and failing tests

ColmBhandal committed:
Unstubbed methods and tests are now passing

Also added some documentation to various methods

ColmBhandal committed:
Added CHANGELOG entry

hctrdev commented on CsharpExtras/Extensions/IEnumerableExtension.cs:
```diff
@@ -36,21 +40,29 @@ public static class IEnumerableExtension
+        public static int FirstIndexOfZeroBased<T>(this IEnumerable<T> values, T value) =>
+            values.FirstIndexOf(value, 0);
+
+        public static int FirstIndexOf<T>(this IEnumerable<T> values, T value, int startIndex)
```
The name `startIndex` is a bit confusing in this context. At first glace I thought it was the index where the search starts (i.e. if you say the start index is 5, it will start looking for a match at position 5 of the array).

ColmBhandal committed:
Renamed startIndex to baseIndex

StartIndex can confuse the user that it's the index at which to start the search
